multus/multus.go

```diff
@@ -215,6 +217,104 @@ func delPlugins(exec invoke.Exec, argIfname string, delegates []*types.DelegateN
 
 	return nil
 }
+func mergeWithResult(srcObj, dstObj cnitypes.Result) (cnitypes.Result, error) {
+	srcObj, err := updateRoutes(srcObj)
+	if err != nil {
+		return nil, logging.Errorf("Routes update failed: %v", err)
+	}
+
+	srcObj, err = fixInterfaces(srcObj)
+	if err != nil {
+		return nil, logging.Errorf("Failed to fix interfaces: %v", err)
+	}
+
+	if dstObj == nil {
+		return srcObj, nil
+	}
+	src, err := current.NewResultFromResult(srcObj)
+	if err != nil {
+		return nil, logging.Errorf("Couldn't convert old result to current version: %v", err)
+	}
+	dst, err := current.NewResultFromResult(dstObj)
+	if err != nil {
+		return nil, logging.Errorf("Couldn't convert old result to current version: %v", err)
+	}
+
+	ifacesLength := len(dst.Interfaces)
+
+	for _, iface := range src.Interfaces {
+		dst.Interfaces = append(dst.Interfaces, iface)
+	}
+	for _, ip := range src.IPs {
+		if ip.Interface != nil && *(ip.Interface) != -1 {
+			ip.Interface = current.Int(*(ip.Interface) + ifacesLength)
+		}
+		dst.IPs = append(dst.IPs, ip)
+	}
+	for _, route := range src.Routes {
+		dst.Routes = append(dst.Routes, route)
+	}
+
+	for _, ns := range src.DNS.Nameservers {
+		dst.DNS.Nameservers = append(dst.DNS.Nameservers, ns)
+	}
+	for _, s := range src.DNS.Search {
+		dst.DNS.Search = append(dst.DNS.Search, s)
+	}
+	for _, opt := range src.DNS.Options {
+		dst.DNS.Options = append(dst.DNS.Options, opt)
+	}
+	// TODO: what about DNS.domain?
+	return dst, nil
+}
+
+// updateRoutes changes nil gateway set in a route to a gateway from IPConfig
+// nil gw in route means default gw from result. When merging results from
+// many results default gw may be set from another CNI network. This may lead to
+// wrong routes.
+func updateRoutes(rObj cnitypes.Result) (cnitypes.Result, error) {
+	result, err := current.NewResultFromResult(rObj)
+	if err != nil {
+		return nil, logging.Errorf("Couldn't convert old result to current version: %v", err)
+	}
+	if len(result.Routes) == 0 {
+		return result, nil
+	}
+
+	var gw net.IP
+	for _, ip := range result.IPs {
+		if ip.Gateway != nil {
+			gw = ip.Gateway
+			break
+		}
+	}
+
+	for _, route := range result.Routes {
+		if route.GW == nil {
+			if gw == nil {
+				return nil, logging.Errorf("Couldn't find gw in result %v", result)
+			}
+			route.GW = gw
+		}
+	}
+	return result, nil
+}
+
+// fixInterfaces fixes bad result returned by CNI plugin
+// some plugins(for example calico) return empty Interfaces list but
+// in IPConfig sets Interface index to 0. In such case it should be -1
+func fixInterfaces(rObj cnitypes.Result) (cnitypes.Result, error) {
+	result, err := current.NewResultFromResult(rObj)
+	if err != nil {
+		return nil, logging.Errorf("Couldn't convert old result to current version: %v", err)
+	}
+	if len(result.Interfaces) == 0 {
+		for _, ip := range result.IPs {
+			ip.Interface = current.Int(-1)
```

Comment:
This should be nil now. See https://github.com/Huawei-PaaS/CNI-Genie/commit/a3fbf3c4dbaff3eba1e23af96d6ae0399ace5458
with -1 containerd will fail

Reply:
Done